test: Enable optional coverage reporting

akshithg opened the pull request:
Add new flag `-c` / `--coverage` to enable optional coverage reporting for all the test types.
- Use `GOFLAGS="-cover -covermode=atomic -coverprofile=...` for go unit tests
- Use the build flag `-cover`, as described in https://go.dev/blog/integration-test-coverage, for the end-to-end go integration tests

aarongable commented on test.sh:
```diff
@@ -104,7 +106,11 @@ With no options passed, runs standard battery of tests (lint, unit, and integrat
     -n, --config-next                     Changes BOULDER_CONFIG_DIR from test/config to test/config-next
     -i, --integration                     Adds integration to the list of tests to run
     -s, --start-py                        Adds start to the list of tests to run
+    -m, --acme                            Adds acme test to the list of tests to run
```
We're not interested in having a "python integration tests only" mode. In fact, we're slowly getting rid of all of the python/chisel-based integration tests, moving everything to Go. Don't bother adding this, imo.

aarongable commented on test.sh:
```diff
@@ -104,7 +106,11 @@ With no options passed, runs standard battery of tests (lint, unit, and integrat
     -g, --generate                        Adds generate to the list of tests to run
+    -c, --coverage                        Enables coverage for tests
+    -d <DIR>, --coverage-directory=<DIR>  Directory to store coverage files in
+                                          Default: test/coverage/<timestamp>
```
Why do you give this a default here, but not in the parser.add_argument() call on line 58 of integration-test.py?

aarongable commented on test.sh:
```diff
@@ -226,6 +240,11 @@ STAGE="unit"
   flush_redis
+
+  if [ "$COVERAGE" == "true" ]; then
+    print_heading "Running unit test with coverage enabled"
```
Remove this debug print

aarongable commented on test.sh:
```diff
@@ -236,11 +255,54 @@ STAGE="integration"
+  fi
+
+  # Add coverage settings if enabled
+  if [ "$COVERAGE" == "true" ]; then
```
```suggestion
  if [ "${COVERAGE}" == "true" ]; then
```

aarongable commented on test.sh:
```diff
@@ -236,11 +255,54 @@ STAGE="integration"
+  python3 test/integration-test.py "${INTEGRATION_ARGS[@]}"
+fi
+
+#
+# ACME tests - these are the chisel tests separated from the main integration
+# tests which also includes go tests.
+#
+STAGE="acme"
+if [[ "${RUN[@]}" =~ "$STAGE" ]] ; then
+  print_heading "Running ACME Tests"
+  flush_redis
+
+  # Set up test parameters
+  ACME_ARGS=("--chisel")
+
+  # Add coverage settings if enabled
+  if [ "$COVERAGE" == "true" ]; then
+    print_heading "Running go acme test with coverage enabled"
+    ACME_ARGS+=("--coverage" "--coveragedir=${COVERAGE_DIR}")
   fi
+
+  # Add any filters
+  ACME_ARGS+=("${FILTER[@]}")
+
+  # Run the acme tests with all collected arguments
+  python3 test/integration-test.py "${ACME_ARGS[@]}"
```
As noted above, I don't think we want or need this mode.

akshithg replied:
I've now removed this mode.

aarongable commented on test/integration-test.py:
```diff
@@ -64,39 +66,48 @@ def main():
     # testing (without having to implement its own busy-wait loop)
     parser.add_argument('--custom', metavar="CMD", help="run custom command")
-    parser.set_defaults(run_chisel=False, test_case_filter="", skip_setup=False)
+    parser.set_defaults(run_chisel=False, test_case_filter="", skip_setup=False, coverage=False, coverage_dir=None)
     args = parser.parse_args()
 
+    if args.coverage and not args.coverage_dir:
+        raise(Exception("must specify --coveragedir when using --coverage"))
```
As noted above, you could easily have a default coveragedir here. Why don't you?

aarongable commented on test/integration-test.py:
```diff
@@ -116,6 +127,7 @@ def main():
     exit_status = 0
 
 def run_chisel(test_case_filter):
+    print("Running ACME tests")
```
Remove debugging print

aarongable commented on test/startservers.py:
```diff
@@ -185,26 +185,36 @@ def _service_toposort(services):
 
+    if coverage:
+        go_build_flags += ' -cover' # https://go.dev/blog/integration-test-coverage
+        print("Building with flags: %s" % go_build_flags)
```
Remove debugging print

aarongable commented on test/startservers.py:
```diff
@@ -185,26 +185,36 @@ def _service_toposort(services):
     e.setdefault("GORACE", "halt_on_error=1")
+    if coverage_dir:
+        abs_coverage_dir = os.path.abspath(coverage_dir)
+        print("Running with coverage in %s" % abs_coverage_dir)
```
Remove debugging print

aarongable reviewed (commented):
Can you speak more to your motivation for adding coverage support, especially to the integration tests? We don't believe in using coverage as a metric for code quality; we find that it often drive perverse incentives and leads to worse code and worse tests just as often as it leads to improvements. I think that adding a `./t.sh -uc` mode for coverage reports on the unit tests makes a lost of sense, as it can be useful during one's own development cycle, but the rest of this PR feels extraneous to me.

I've left small technical comments below, but mostly I think my feedback is this: pare this down to just a new `-c` flag in test.sh which adds coverage output for the Go unittests, and we'll gladly accept it.

aarongable commented:
Thanks for the explanation! I think that's a fantastic goal, and I'm totally in favor of adding the ability to easily get coverage reports from our unit and integration tests.

One minor concern: tracking Boulder code-level coverage is a very different beast from tracking ACME protocol-level coverage. Huge portions of the Boulder codebase are dedicated to handling the difference between a dns-01 challenge and a tls-alpn-01 challenge, or the difference between a DNS identifier and an IP identifier, or the difference between an RSA pubkey and an ECDSA pubkey... but none of those differences appear anywhere in the ACME state diagrams. So I'm not totally sure how tracking code-level coverage helps with this effort. But that's relatively minor, and I'm still in favor of easy-to-access coverage reports as a diagnostic and development tool.

That said, I'd like this to take a slightly different shape: I'd like to only track coverage from our [_Go-based_ integration tests](https://github.com/letsencrypt/boulder/tree/main/test/integration). We are actively working on getting rid of the python-based chisel tests (see https://github.com/letsencrypt/boulder/pull/8258, https://github.com/letsencrypt/boulder/pull/8085, and https://github.com/letsencrypt/boulder/pull/8082, for example). If we want to improve protocol-level coverage of the Boulder integration tests, I want all of that improvement to happen within the Go integration tests, not the python ones.

akshithg commented:
Got it --- thanks for clarifying, and I see your point about the difference between code-level and protocol-level coverage. For my purposes, I'm fine using code-level coverage as a _proxy_ in the short term to spot untested flows, and then drilling deeper into protocol-specific coverage metrics separately.

I wasn't aware of the active effort to retire the Python-based chisel tests. That makes sense, and I'm happy to align with that direction.

For this PR, I can pare it back so that:

- Coverage support applies to Go-based unit tests and Go-based integration tests only.
- Chisel-specific coverage logic is removed.

That way, we still gain the diagnostic capability you're supportive of, while keeping future improvements in line with the Go-only integration testing plan.

akshithg commented on test/integration-test.py:
```diff
@@ -55,6 +53,11 @@ def main():
                         help="run integration tests using chisel")
+    parser.add_argument('--coverage', dest="coverage", action="store_true",
+                        help="run integration tests with coverage")
+    parser.add_argument('--coveragedir', dest="coverage_dir", action="store",
+                        default=f"test/coverage/{datetime.datetime.now().strftime('%Y-%m-%d_%H-%M-%S')}",
```
`--coveragedir` now has a default value similar to the one used in `test.sh`

akshithg commented on test.sh:
```diff
@@ -226,6 +238,10 @@ STAGE="unit"
   flush_redis
+
+  if [ "$COVERAGE" == "true" ]; then
+    export GOFLAGS="-cover -covermode=atomic -coverprofile=${COVERAGE_DIR}/unit.coverprofile -coverpkg=./..."
```
That is a good idea. I'll update `-coverpkg` to respect `UNIT_PACKAGES`

aarongable commented:
LGTM, needs a second review + approval to merge.

akshithg commented:
Hi @jsha,
I hope you’re doing well. I wanted to kindly follow up on this PR. @aarongable has already reviewed and approved the changes after the revisions, and I’d greatly appreciate it if you could take a look when you have a moment and provide the second review/approval so we can get this merged.

Thank you very much!

akshithg commented:
Hi @beautifulentropy, thanks for the update!
I've fixed the merge conflicts, rebased, and squashed the commits. The PR should now be ready for review. Please let me know if there's anything else I should adjust.
